File: frontend/src/api/tenants.ts
import { api } from "./client";

export type Tenant = { id: string; name: string };
export type School = { id: string; tenant_id: string; name: string };

export const tenantsApi = {
  list: () => api<Tenant[]>("/tenants", { method: "GET", auth: true }),
  create: (name: string) =>
    api<Tenant>("/tenants", { body: { name }, auth: true }),
  patch: (id: string, name: string) =>
    api<Tenant>(`/tenants/${id}`, { method: "PATCH", body: { name }, auth: true }),
  remove: (id: string) =>
    api<void>(`/tenants/${id}`, { method: "DELETE", auth: true }),
};

export const schoolsApi = {
  list: () => api<School[]>("/schools", { method: "GET", auth: true }),
  create: (tenant_id: string, name: string) =>
    api<School>("/schools", { body: { tenant_id, name }, auth: true }),
  patch: (id: string, name: string) =>
    api<School>(`/schools/${id}`, { method: "PATCH", body: { name }, auth: true }),
  remove: (id: string) =>
    api<void>(`/schools/${id}`, { method: "DELETE", auth: true }),
};

File: backend/app/api/v1/routes/tenants.py
from fastapi import APIRouter, Depends, Response, status
from pydantic import BaseModel
from sqlalchemy.orm import Session

from app.api.deps import require_role
from app.db.session import get_db
from app.model.user import User
from app.schema.tenant import TenantCreate, TenantRead
from app.service.tenant import TenantService

router = APIRouter(prefix="/tenants", tags=["tenants"])


class TenantPatch(BaseModel):
    name: str | None = None


@router.get("", response_model=list[TenantRead])
def list_tenants(
    _: User = Depends(require_role("admin")), db: Session = Depends(get_db)
):
    return TenantService(db).list_all()


@router.post("", response_model=TenantRead, status_code=status.HTTP_201_CREATED)
def create_tenant(
    payload: TenantCreate,
    _: User = Depends(require_role("admin")),
    db: Session = Depends(get_db),
):
    return TenantService(db).create(payload.name)


@router.patch("/{tenant_id}", response_model=TenantRead)
def patch_tenant(
    tenant_id: str,
    payload: TenantPatch,
    _: User = Depends(require_role("admin")),
    db: Session = Depends(get_db),
):
    return TenantService(db).patch(tenant_id, payload.model_dump(exclude_unset=True))


@router.delete("/{tenant_id}", status_code=status.HTTP_204_NO_CONTENT)
def delete_tenant(
    tenant_id: str,
    _: User = Depends(require_role("admin")),
    db: Session = Depends(get_db),
):
    TenantService(db).delete(tenant_id)
    return Response(status_code=status.HTTP_204_NO_CONTENT)